File: backend-express/src/lib/encryption/token.ts
import { env } from "@/config";
import {
    DecodedTokenPayload,
    EncodedToken,
    TokenPayload,
    UserRead,
} from "@/models/schemas";
import { ApiError, HttpStatus } from "@/types";

import { decodePayload, encodePayload } from "./helpers";

export const decodeToken = (encoded: string): DecodedTokenPayload => {
    const decoded = decodePayload(encoded);
    if (typeof decoded === "string") {
        throw new ApiError(HttpStatus.BAD_REQUEST, "Invalid token payload");
    }
    return decoded as DecodedTokenPayload;
};

export const createToken = (user: UserRead): EncodedToken => {
    const payload: TokenPayload = {
        userId: user.id,
        email: user.email,
    };
    const access_token = encodePayload(payload);
    return {
        access_token,
        token_type: "bearer",
        email: user.email,
        userId: user.id,
        expires_in: env.JWT_EXPIRATION,
    };
};

File: backend-gin/internal/lib/encryption/token.go
package encryption

import (
	"backend/internal/config"
	"backend/internal/lib/utils"
	"backend/internal/models/schemas"
	"fmt"
	"time"
)

func DecodeEncodedToken(encoded string) (schemas.TokenPayload, error) {
	var zero schemas.TokenPayload

	decoded, err := utils.DecodePayload(encoded, config.Env.SecretKey)
	if err != nil {
		return zero, fmt.Errorf(
			"token %s not valid, could not decode it: %w", encoded, err,
		)
	}

	userIdRaw, found := decoded["userId"]
	if !found {
		return zero, fmt.Errorf("token %s not valid, no userId found", encoded)
	}
	userId, ok := userIdRaw.(string)
	if !ok {
		return zero, fmt.Errorf(
			"token %s not valid, userId %s not valid", encoded, userIdRaw,
		)
	}

	emailRaw, found := decoded["email"]
	if !found {
		return zero, fmt.Errorf("token %s not valid, no email found", encoded)
	}
	email, ok := emailRaw.(string)
	if !ok {
		return zero, fmt.Errorf(
			"token %s not valid, email %s not valid", encoded, emailRaw,
		)
	}

	return schemas.TokenPayload{
		UserId: userId, Email: email,
	}, nil
}

func CreateToken(id string, email string) (schemas.EncodedToken, error) {
	payload := map[string]any{"userId": id, "email": email}
	acccessToken, err := utils.EncodePayload(
		payload, config.Env.SecretKey, time.Duration(config.Env.JWTExpiration)*time.Second,
	)
	if err != nil {
		var zero schemas.EncodedToken
		return zero, fmt.Errorf("could not encode payload for user %s", id)
	}

	return schemas.EncodedToken{
		AccessToken: acccessToken,
		TokenType:   "bearer",
		Email:       email,
		UserId:      id,
		ExpiresIn:   config.Env.JWTExpiration,
	}, nil
}
